scipy/optimize/_constraints.py

```diff
@@ -0,0 +1,620 @@
+from __future__ import division, print_function, absolute_import
+import numpy as np
+import scipy.sparse as spc
+
+
+__all__ = ['CanonicalConstraint',
+           'NonlinearConstraint',
+           'LinearConstraint',
+           'BoxConstraint',
+           'check_sparsity',
+           'parse_constraint',
+           'reinforce_box_constraints',
+           'empty_canonical_constraint',
+           'generate_lagrangian_hessian',
+           'concatenate_canonical_constraints']
+
+
+class CanonicalConstraint:
+    """Canonical constraint
+
+    Constraint of the form:
+        constr_ineq(x) <= 0
+        constr_eq(x) = 0
+
+    Parameters
+    ----------
+    n_ineq : int
```

Comment:
I am doing it for efficiency. Furthermore, the ``CanonicalConstraint`` class is more like a low level interface, not directly available to the user. That is why their parameters are not user friendly and I do not check them.

The classes that should be available to the user are:
    - ``BoxConstraint``
    - ``LinearConstraint``
    - ``NonlinearConstraint``